nntrainer/models/neuralnet.cpp

```diff
@@ -771,46 +779,49 @@ int NeuralNetwork::train_run(std::function<bool(void *userdata)> stop_cb) {
       auto const &inputs = iteration.getInputsRef();
       model_graph.setInputsLabels(inputs, labels);
 
-      on_iteration_fetch(stat, *buffer);
+      on_iteration_fetch(stat, *buffer, iter_num);
       on_iteration_update_stat(stat, outputs, labels);
     }
     future_iq.get();
-    on_epoch_end(stat, *buffer);
+    on_epoch_end(stat, *buffer, iter_num);
 
     if (stat.num_iterations == 0) {
       throw std::runtime_error("No data came while buffer ran");
     }
-
-    return stat;
   };
 
-  auto train_for_iteration = [this, stop_cb](RunStats &stat,
-                                             DataBuffer &buffer) {
-    forwarding(true, stop_cb);
-    backwarding(iter++, stop_cb);
-
-    // To avoid unconsidered memory leak, we need to clear the cache
-    model_graph.flushCache();
-
-    if (!stop_cb(nullptr)) {
-      std::cout << "#" << epoch_idx << "/" << getEpochs();
-      ml_logi("# %d / %d", epoch_idx, getEpochs());
-      auto loss = getLoss();
-      buffer.displayProgress(stat.num_iterations, loss);
-    }
-  };
+  auto train_for_iteration =
+    [this, stop_cb, user_data](ml::train::RunStats &stat, DataBuffer &buffer,
+                               unsigned int &iter_num) {
+      forwarding(true, stop_cb, user_data);
+      backwarding(iter++, stop_cb, user_data);
+
+      // To avoid unconsidered memory leak, we need to clear the cache
+      model_graph.flushCache();
+
+      if (!stop_cb(user_data)) {
+        ++iter_num;
+        std::cout << "iteration: " << iter_num;
+        ml_logi("iteration: %d", iter_num);
+        auto loss = getLoss();
+        buffer.displayProgress(stat.num_iterations, loss);
+      }
+    };
 
-  auto update_train_stat = [this](RunStats &stat,
+  auto update_train_stat = [this](ml::train::RunStats &stat,
                                   const std::vector<Tensor> &outputs,
                                   const std::vector<Tensor> &labels) {
     stat.loss += getLoss();
     stat.num_iterations++;
   };
 
-  auto train_epoch_end = [this, stop_cb](RunStats &stat, DataBuffer &buffer) {
+  auto train_epoch_end = [this, stop_cb, user_data](ml::train::RunStats &stat,
+                                                    DataBuffer &buffer,
+                                                    unsigned int &iter_num) {
     stat.loss /= static_cast<float>(stat.num_iterations);
     auto &save_path = std::get<props::SavePath>(model_flex_props);
-    if (!stop_cb(nullptr)) {
+    if (!stop_cb(user_data)) {
+      iter_num = 0;
```

Comment:
on stop, current iteration information will be deleted. Is it intended ?
I think it's better to maintain current status until user reset/reconfigure training status.

Reply:
1. Line 824 was meant to be called only if the stop_cb is return false which means normal termination. If so at the end of every epoch iter_num was initialize to 0.
2. This code is removed and initialize iter_num at the beginning of every epoch